📖 Runtime SDK: extend documentation & update proposals accordingly (#6756)
kubernetes-sigs/cluster-api: docs/book/src/tasks/experimental-features/runtime-sdk/implement-extensions.md

a Runtime Hook, you are required to implement a Runtime Extension handling requests according to the
OpenAPI specification for the Runtime Hook you are interested in.

Runtime Extensions by design are very powerful and flexible, however given that with great power comes
great responsibility, a few key consideration should always be kept in mind (more details in the following paragraphs):

- Runtime Extension are components that should be designed, written and deployed with great caution given that they
  can affect the proper functioning of the Cluster API runtime.
- Cluster administrators should carefully vet any Runtime Extension registration, thus preventing malicious components
  from being added to the system.

Please note that following similar practices is already commonly accepted in the Kubernetes ecosystem for
Kubernetes API server admission webhooks, and Runtime Extensions share the same foundation and most of the same
considerations/concerns apply.

## Implementation

As mentioned above as a developer building systems on top of Cluster API, if you want to hook in the Cluster’s lifecycle via a

> Runtime Extension, you have to implement an HTTP server handling a discovery request and a set of

Author: I did it, but I'm not sure if folks usually say HTTP or HTTPS server even if we only support https

> Runtime Extension, you have to implement an HTTP server handling a discovery request and a set of

Reviewer: I think it might be best to be explicit here in terms of expectations - agreed that HTTP in this context usually means the protocal +/- S though 😄

additional requests according to the OpenAPI specification for the Runtime Hook you are interested in.

The following shows a minimal example of a Runtime Extension server implementation:

> ```go

Author: Thank you :)

Kubernetes API server does for different API resources. The exact format of those paths is handled by the `server` 
package automatically.

There is an additional `Discovery` endpoint which is automatically handled by the `Server`. The `Discovery` endpoint
returns a list of the implemented Runtime Hooks to inform Cluster API which Runtime Hooks are implemented by this 
Runtime Extension server.

Please note that Cluster API only enforces request and response types as defined by a Runtime Hook version,
and developers are fully responsible for all other elements of the design of a Runtime Extension implementation,
including:

- To choose which programming language to use; please note that for sake of this proposal Golang is the language
  of choice, and we are not planning to test/provide tooling/libraries for other languages. Nevertheless, given that
  we rely on Open API and plain HTTP(s) calls, other languages should just work but support will be provided at
  best effort.
- To choose if to have a dedicated HTTP server(s) for Runtime Extensions only or if to use the HTTP server for other
  purposes as well (e.g. to serve a metric endpoint).

> In case the Runtime Extension is developed in Golang, the implementer can benefit from the following packages (exported by

Reviewer: ```suggestion
When using Golang the Runtime Extension is developer can benefit from the following packages (exported by
```

`sigs.k8s.io/cluster-api`) as shown in the example above:

- `exp/runtime/hooks/api/v1alpha1` contains the Runtime Hook Golang API types, which are also used to generate the OpenAPI specification.
- `exp/runtime/catalog` provides the `Catalog` object to register Runtime Hook definitions. The `Catalog` is then used by the `server` 

>   package to handle requests. `Catalog` is similar to the `runtime.Scheme` of the `k8s.io/apimachinery/pkg/runtime`
>   package just for Runtime Hooks.

Reviewer: ```suggestion
  package to handle requests. `Catalog` is similar to the `runtime.Scheme` of the `k8s.io/apimachinery/pkg/runtime`
  package, but is designed for Runtime Hook registration.
```

* `./exp/runtime/server`: The `server` package provides a `Server` object which makes it easy to implement
  a Runtime Extension server. The `Server` will automatically handle tasks like Marshal/Unmarshal. A Runtime Extension 
  developer only has to implement a strongly typed function that contains the actual extension implementation.

## Guidelines

While writing the actual code of the Runtime Extension a set of important guidelines must be considered:

### Timeouts

> Runtime Extension processing adds to network request latency, they should run as quickly as possible
> (typically in milliseconds); Cluster Administrator will be allowed to configure how long the Cluster API Runtime
> should wait for a Runtime Extension to respond before treating the call as a failure (max 10s).
> 

Reviewer: ```suggestion
Runtime Extension processing adds to network request latency, they should run as quickly as possible
(typically in milliseconds); Cluster Administrators will be allowed to configure how long the Cluster API Runtime
should wait for a Runtime Extension to respond before treating the call as a failure (max 10s).

```

### Availability

Runtime Extension failure could result in errors in handling the workload clusters lifecycle, and so the implementation

> should be robust, have proper error handling, avoid panics, etc.. It will be allowed to set up
> failure policies preventing a Runtime Extension failure to have negative effects on the Cluster API Runtime, but
> this option can’t be used in all use cases (see [Error Management](#error-management)).

Reviewer: ```suggestion
should be robust, have proper error handling, avoid panics, etc.. Failure policies can be set up to mitigate the negative impact of a Runtime Extension on the Cluster API Runtime, but
this option can’t be used in all use cases (see [Error Management](#error-management)).
```

### Blocking Hooks

> A Runtime Hook can be defined as "blocking", e.g. the `BeforeClusterUpgrade` hook allows a Runtime Extension
> to prevent the upgrade to start. A Runtime Extension registered for the `BeforeClusterUpgrade` hook can block 
> by retuning a non-zero `retryAfterSeconds` value. Following consideration apply:
> 

Reviewer: ```suggestion
A Runtime Hook can be defined as "blocking" -  e.g. the `BeforeClusterUpgrade` hook allows a Runtime Extension
to prevent the upgrade process from starting. A Runtime Extension registered for the `BeforeClusterUpgrade` hook can block 
by retuning a non-zero `retryAfterSeconds` value. Following consideration apply:

```

- The system might decide to retry the same Runtime Extension even before the `retryAfterSeconds` period expires,
  e.g. due to other changes in the Cluster, so `retryAfterSeconds` should be considered as an approximate maximum
  time before the next reconcile.
- If there is more than one Runtime Extension registered for the same Runtime Hook and more than one returns
  `retryAfterSeconds`, the shortest non-zero value will be used.
- If there is more than one Runtime Extension registered for the same Runtime Hook and at least one returns
  `retryAfterSeconds`, all Runtime Extension will be re-tried.

Detailed description of what "blocking" means for each specific Runtime Hooks will be documented case by case.

### Side Effects

> It is recommended that Runtime Extensions should avoid side effects if possible, which means to operate only on
> the content of the request sent to them, and not make out-of-band changes.
> If side effects are required, rules defined in the following paragraphs apply.

Reviewer: ```suggestion
It is recommended that Runtime Extensions should avoid side effects if possible, which means they should operate only on
the content of the request sent to them, and not make out-of-band changes.
If side effects are required, rules defined in the following paragraphs apply.
```

# Implementing Runtime Extensions

<aside class="note warning">

Please note Runtime SDK is an advanced feature. If implemented incorrectly, a failing Runtime Extension can severely impact the Cluster API runtime.

> </aside>

Author: Thx, done

great responsibility, a few key consideration should always be kept in mind (more details in the following sections):

- Runtime Extensions are components that should be designed, written and deployed with great caution given that they
  can affect the proper functioning of the Cluster API runtime.
- Cluster administrators should carefully vet any Runtime Extension registration, thus preventing malicious components
  from being added to the system.

Please note that following similar practices is already commonly accepted in the Kubernetes ecosystem for
Kubernetes API server admission webhooks. Runtime Extensions share the same foundation and most of the same
considerations/concerns apply.

## Implementation

As mentioned above as a developer building systems on top of Cluster API, if you want to hook in the Cluster’s 
lifecycle via a Runtime Extension, you have to implement an HTTPS server handling a discovery request and a set 
of additional requests according to the OpenAPI specification for the Runtime Hook you are interested in.

The following shows a minimal example of a Runtime Extension server implementation:

```go

> var (

Author: Intentionally didn't add it because it makes the example even bigger, but added it now

		klog.Infof("Profiler listening for requests at %s", profilerAddress)
		go func() {
			klog.Info(http.ListenAndServe(profilerAddress, nil))
		}()
	}

	ctx := ctrl.SetupSignalHandler()

	webhookServer, err := server.NewServer(server.Options{
		Catalog: catalog,
		Port:    webhookPort,
		CertDir: webhookCertDir,
	})
	if err != nil {
		setupLog.Error(err, "error creating webhook server")
		os.Exit(1)
	}

	if err := webhookServer.AddExtensionHandler(server.ExtensionHandler{
		Hook:           runtimehooksv1.BeforeClusterCreate,
		Name:           "before-cluster-create",
		HandlerFunc:    DoBeforeClusterCreate,
		TimeoutSeconds: pointer.Int32(5),
		FailurePolicy:  toPtr(runtimehooksv1.FailurePolicyFail),
	}); err != nil {
		setupLog.Error(err, "error adding handler")
		os.Exit(1)
	}
	
	if err := webhookServer.AddExtensionHandler(server.ExtensionHandler{
		Hook:           runtimehooksv1.DoBeforeClusterUpgrade,
		Name:           "before-cluster-upgrade",
		HandlerFunc:    DoBeforeClusterUpgrade,
		TimeoutSeconds: pointer.Int32(5),
		FailurePolicy:  toPtr(runtimehooksv1.FailurePolicyFail),
	}); err != nil {
		setupLog.Error(err, "error adding handler")
		os.Exit(1)
	}

> 	setupLog.Info("Starting Runtime Extension server")

Author: I would prefer not making the example even bigger, that's why I added the explanation below. I would prefer not also adding it inline.

	log := ctrl.LoggerFrom(ctx)
	log.Info("BeforeClusterCreate is called")
	// Your implementation
}

func DoBeforeClusterUpgrade(ctx context.Context, request *runtimehooksv1.BeforeClusterUpgradeRequest, response *runtimehooksv1.BeforeClusterUpgradeResponse) {
	log := ctrl.LoggerFrom(ctx)
	log.Info("BeforeClusterUpgrade is called")
	// Your implementation
}

func toPtr(f runtimehooksv1.FailurePolicy) *runtimehooksv1.FailurePolicy {
	return &f
}
```

For a full example see our [test extension](https://github.com/kubernetes-sigs/cluster-api/tree/main/test/extension). 

Please note that a Runtime Extension server can serve multiple Runtime Extensions (in the example above 
`BeforeClusterCreate` and `BeforeClusterUpgrade`) at the same time. Each of them are handled at a different path, like the 
Kubernetes API server does for different API resources. The exact format of those paths is handled by the server
automatically.

There is an additional `Discovery` endpoint which is automatically served by the `Server`. The `Discovery` endpoint
returns a list of the added extension handlers to inform Cluster API which Runtime Hooks are implemented by this 

> Runtime Extension server.

Author: I changed the initial wording a bit, but I think "Runtime Hooks are available" is not the right wording 

(give that the hooks are what we implemented in the core controllers)

Please note that Cluster API only enforces the correct request and response types as defined by a Runtime Hook version.
Developers are fully responsible for all other elements of the design of a Runtime Extension implementation, including:

- To choose which programming language to use; please note that for sake of this proposal Golang is the language
  of choice, and we are not planning to test or provide tooling and libraries for other languages. Nevertheless, given that
  we rely on Open API and plain HTTPS calls, other languages should just work but support will be provided at best effort.
- To choose if a dedicated HTTPS server is used for Runtime Extensions only or if the HTTPS server is used for other

>   purposes as well (e.g. to serve a metric endpoint).

Author: Ended up with a mixture of both.
I was thinking about just dropping it to be honest :)

> When using Golang the Runtime Extension developer can benefit from the following packages (exported by `sigs.k8s.io/cluster-api`)

Author: Also changed the initial wording to something else :P

`sigs.k8s.io/cluster-api` module) as shown in the example above:

- `exp/runtime/hooks/api/v1alpha1` contains the Runtime Hook Golang API types, which are also used to generate the 
  OpenAPI specification.
- `exp/runtime/catalog` provides the `Catalog` object to register Runtime Hook definitions. The `Catalog` is then 
  used by the `server` package to handle requests. `Catalog` is similar to the `runtime.Scheme` of the 
  `k8s.io/apimachinery/pkg/runtime` package, but it is designed to store Runtime Hook registrations.
- `./exp/runtime/server` provides a `Server` object which makes it easy to implement a Runtime Extension server.
  The `Server` will automatically handle tasks like Marshalling/Unmarshalling requests and responses. A Runtime 
  Extension developer only has to implement a strongly typed function that contains the actual logic.

## Guidelines

While writing a Runtime Extension the following important guidelines must be considered:

### Timeouts

Runtime Extension processing adds to reconcile durations of Cluster API controllers. They should respond to requests 

> as quickly as possible, typically in milliseconds. Cluster Administrators will be allowed to configure how long the 

Author: Good finding. Yeah that sentence was from where it was still a part of ExtensionConfig

> Cluster API Runtime should wait for a Runtime Extension to respond before treating the call as a failure (max is 10s).

Author: Thx!

> Cluster API Runtime should wait for a Runtime Extension to respond before treating the call as a failure (max is 10s).

Author: Added something about this

### Availability

Runtime Extension failure could result in errors in handling the workload clusters lifecycle, and so the implementation
should be robust, have proper error handling, avoid panics, etc.. . Failure policies can be set up to mitigate the 
negative impact of a Runtime Extension on the Cluster API Runtime, but this option can’t be used in all cases 
(see [Error Management](#error-management)).

> ### Blocking Hooks

Author: We have some language around that in the error management section.

I would keep it as is for now. We can improve on it later if there's demand

A Runtime Hook can be defined as "blocking" -  e.g. the `BeforeClusterUpgrade` hook allows a Runtime Extension
to prevent the upgrade from starting. A Runtime Extension registered for the `BeforeClusterUpgrade` hook 
can block by returning a non-zero `retryAfterSeconds` value. Following consideration apply:

- The system might decide to retry the same Runtime Extension even before the `retryAfterSeconds` period expires,
  e.g. due to other changes in the Cluster, so `retryAfterSeconds` should be considered as an approximate maximum
  time before the next reconcile.
- If there is more than one Runtime Extension registered for the same Runtime Hook and more than one returns
  `retryAfterSeconds`, the shortest non-zero value will be used.
- If there is more than one Runtime Extension registered for the same Runtime Hook and at least one returns
  `retryAfterSeconds`, all Runtime Extension will be called again.

Detailed description of what "blocking" means for each specific Runtime Hooks is documented case by case 

> in the hook-specific implementation documentation.

Author: Added it via e.g.. Don't want to link lifecycle hooks as the "one" hook-specific implementation documentation

	log.Info("BeforeClusterUpgrade is called")
	// Your implementation
}

func toPtr(f runtimehooksv1.FailurePolicy) *runtimehooksv1.FailurePolicy {
	return &f
}
```

For a full example see our [test extension](https://github.com/kubernetes-sigs/cluster-api/tree/main/test/extension). 

Please note that a Runtime Extension server can serve multiple Runtime Hooks (in the example above 
`BeforeClusterCreate` and `BeforeClusterUpgrade`) at the same time. Each of them are handled at a different path, like the 
Kubernetes API server does for different API resources. The exact format of those paths is handled by the server
automatically in accordance to the OpenAPI specification of the Runtime Hooks.

There is an additional `Discovery` endpoint which is automatically served by the `Server`. The `Discovery` endpoint
returns a list of extension handlers to inform Cluster API which Runtime Hooks are implemented by this 
Runtime Extension server.

Please note that Cluster API is only able to enforce the correct request and response types as defined by a Runtime Hook version.
Developers are fully responsible for all other elements of the design of a Runtime Extension implementation, including:

> - To choose which programming language to use; please note that for sake of this proposal Golang is the language

Reviewer 2: nit
```suggestion
- To choose which programming language to use; please note that for sake of this documentation Golang is the language
```

### Error Management

In case a Runtime Extension returns an error, the error will be handled according to the corresponding failure policy
defined in the response of the Discovery call.

If the failure policy is `Ignore` the error is going to be recorded in the controller's logs, but the processing 
will continue. However we recognize that this failure policy cannot be used in most of the use cases because Runtime 
Extension implementers want to ensure that the task implemented by an extension is completed before continuing with 
the cluster's lifecycle.

If instead the failure policy is `Fail` the system will retry the operation until it passes. The following general 
considerations apply:

- It is the responsibility of Cluster API components to surface Runtime Extension errors using conditions.
- Operations will be retried with an exponential backoff or whenever the state of a Cluster changes (we are going to rely
  on controller runtime exponential backoff/watches).
- If the operation is defined as "blocking", the error is going to block a lifecycle transition, e.g. returning a non-zero 

>   `retryAfterSeconds` on a Runtime Extension for the `BeforeClusterUpgrade` hook is going to block the Cluster upgrade.

Author: A non-zero retryAfterSeconds will only block the upgrade, an error will fail the entire reconcile.

The blocking behavior we're referring to is specifically the behavior of a non-zero retryAfterSeconds.

Returning errors will fail the entire reconciles for all hooks including non-blocking hooks

>   `retryAfterSeconds` on a Runtime Extension for the `BeforeClusterUpgrade` hook is going to block the Cluster upgrade.

Reviewer 2: Ok, got it, but what I find confusing is that we are in the error management paragraph, in a sentence that starts with "If the operation is defined as "blocking", the error is going to block a lifecycle transition, e.g.", but then the example talks about retryAfterSeconds
This seems inconsistent to me...

>   `retryAfterSeconds` on a Runtime Extension for the `BeforeClusterUpgrade` hook is going to block the Cluster upgrade.

Author: Ah good point. I'll drop the part with blocking. I think then it makes sense and fits here

>   `retryAfterSeconds` on a Runtime Extension for the `BeforeClusterUpgrade` hook is going to block the Cluster upgrade.

Author: Re-reading this. I'll drop this point given that the point above explains the exponential backoff and the whole listing is about the error case
